### PR TITLE
scryptWebWorker

### PR DESCRIPTION
**Pull request by soshochang**
-add scryptWebWorker so that scrypt key derivation does not freeze loading screen ui, modify associated login and derivePasswordHash methods to use it

### CODE REVIEW

File: src/background/controllers/cryptoController.ts [hunk 1]
```diff
@@ -64,26 +63,36 @@ export default class CryptoController extends IController {
...
   */
-  public derivePasswordHash = async (password: string): Promise<any> => {
-    return new Promise((resolve: any, reject: any) => {
-      setTimeout(() => {
-        try {
-          if (!this.appSalt) {
-            throw Error('appSalt should not be empty');
-          }
+  public derivePasswordHash = async (password: string) => {
+    setTimeout(() => {
```
Comment (dwalintukan) [inline src/background/controllers/cryptoController.ts]: should be able to remove this timeout

File: src/background/controllers/cryptoController.ts [hunk 2]
```diff
@@ -64,26 +63,36 @@ export default class CryptoController extends IController {
...
+      if (!this.appSalt) {
+        throw Error('appSalt should not be empty');
+      }
+      this.callScryptWebWorker(password, this.appSalt);
```
Comment (dwalintukan) [inline src/background/controllers/cryptoController.ts]: inline the function here

File: src/background/controllers/cryptoController.ts [hunk 3]
```diff
@@ -64,26 +63,36 @@ export default class CryptoController extends IController {
...
+    }, 100);
+  }
 
-          const saltBuffer = Buffer.from(this.appSalt!);
-          const { N, r, p } = CryptoController.SCRYPT_PARAMS_PW;
-          const derivedKey = scrypt(password, saltBuffer, N, r, p, 64);
-          this.passwordHash = derivedKey.toString('hex');
+  public callScryptWebWorker = (password: string, salt: Uint8Array) => {
+    let sWW;
```
Comment (dwalintukan) [inline src/background/controllers/cryptoController.ts]: nit: `sww` or `worker`

File: src/background/controllers/cryptoController.ts [hunk 4]
```diff
@@ -64,26 +63,36 @@ export default class CryptoController extends IController {
...
+      * Create a web worker for the scrypt key derivation, so that it doesn't freeze the loading screen ui.
+      * File path relative to post bundling of webpack. worker-loader node module did not work for me,
+      * possibly a compatibility issue with chrome.
+      */
+      sWW = new Worker('./scryptworker.js');
 
-          resolve();
-        } catch (err) {
-          reject(err);
+      sWW.postMessage({password, salt, scryptParams: CryptoController.SCRYPT_PARAMS_PW});
```
Comment (dwalintukan) [inline src/background/controllers/cryptoController.ts]: spaces at beginning and end of { }

File: src/background/controllers/cryptoController.ts [hunk 5]
```diff
@@ -64,26 +63,36 @@ export default class CryptoController extends IController {
...
+
+      sWW.onmessage = (e) => {
+        if (e.data.err) {
+          throw new Error('scrypt failed to calculate derivedKey');
```
Comment (dwalintukan) [inline src/background/controllers/cryptoController.ts]: nit: `throw Error`

File: webpack.base.config.js [hunk 6]
```diff
@@ -38,6 +38,7 @@ module.exports = {
     contentscript: './src/contentscript/index.ts',
     inpage: './src/inpage/index.ts',
     popup: './src/popup/index.tsx',
+    scryptworker: './src/background/workers/scryptworker.js',
```
Comment (dwalintukan) [inline webpack.base.config.js]: can we access this by defining it via `web_accessible_resources`? from what i gathered thats generally how the extension should access files from the build files.

File: src/background/controllers/cryptoController.ts (hunk 2 above)
Comment (soshochang) [reply src/background/controllers/cryptoController.ts]: do you mean to turn it into a ternary? I I think it is kind of hard to read.  It doesn't seem like you can just throw in a ternary

this.appSalt ? this.callScryptWebWorker(password, this.appSalt) 
        : function() { throw Error('appSalt should not be empty') }()

File: webpack.base.config.js (hunk 6 above)
Comment (soshochang) [reply webpack.base.config.js]: just tried it, doesn't work, I get an uncaught syntax error for the import line. Looks like we need webpack to take care of the importing.

File: src/background/controllers/cryptoController.ts (hunk 2 above)
Comment (dwalintukan) [reply src/background/controllers/cryptoController.ts]: Inlining is moving the entire logic of callScryptWebWorker into the derivePasswordHash func.

File: webpack.base.config.js (hunk 6 above)
Comment (dwalintukan) [reply webpack.base.config.js]: Gotcha. Adding it to webpack must allow the node code to be converted.

**Comment by dwalintukan** [comment]
Finish inlining then merge 👍

**Pull request merged**